### PR TITLE
sched/pthread:need check pthread is DETACHED

### PR DESCRIPTION
**Pull request by anjiahao1**
pthread_join need check thread is DETACHED,
Whether to wait according to the result.And,
if a thread is DETACHED,it will not set a new
attr.
Signed-off-by: anjiahao <anjiahao@xiaomi.com>

## Summary
Change thread behavior
## Impact
maybe will impact thread is DETACHED

### CODE REVIEW

File: sched/pthread/pthread_detach.c [hunk 1]
```diff
@@ -96,12 +96,15 @@ int pthread_detach(pthread_t thread)
            * thread exits
            */
 
-          pjoin->detached = true;
+          if (pjoin->detached)
+            {
+              ret = EINVAL;
```
Comment (pkarashchenko) [inline sched/pthread/pthread_detach.c:101]: I think that standard does not define this behavior. Are you trying to bring Linux behavior here? I mean what is the motivation of this change?

Comment (pkarashchenko) [reply sched/pthread/pthread_detach.c:101]: I mean `The behavior is undefined if the value specified by the thread argument to pthread_detach() does not refer to a joinable thread.`

File: sched/pthread/pthread_join.c [hunk 2]
```diff
@@ -152,90 +152,99 @@ int pthread_join(pthread_t thread, FAR pthread_addr_t *pexit_value)
...
-      pjoin->crefs++;
-
-      /* Check if the thread is still running.  If not, then things are
-       * simpler.  There are still race conditions to be concerned with.
-       * For example, there could be multiple threads executing in the
-       * 'else' block below when we enter!
-       */
-
-      if (pjoin->terminated)
+      if (!pjoin->detached)
```
Comment (pkarashchenko) [inline sched/pthread/pthread_join.c:155]: I see this a bit inconsistent. You added code to return `EINVAL` in `pthread_detach` is case if thread is already detached, but here you do not set `ret` to `EINVAL` is case if trying to join a detached thread. But according to spec `The behavior is undefined if the value specified by the thread argument to pthread_join() does not refer to a joinable thread.`, so it is the same `undefined` and for `pthread_detach`.

Comment (anjiahao1) [reply sched/pthread/pthread_join.c:155]: i use `sudo apt-get install  manpages-posix manpages-posix-dev ` and `man 3posix ptread_detach` ，that suggest `If an implementation detects that the value specified by the thread argument to  pthread_detach()  does  not refer to a joinable thread, it is recommended that the function should fail and report an [EINVAL] error`

Comment (anjiahao1) [reply sched/pthread/pthread_join.c:155]: and,i use `phtread_jion` to wait a `detached thread` it also will wait. `phtread_jion` wait a `detached thread` should return now.

Comment (pkarashchenko) [reply sched/pthread/pthread_join.c:155]: Return now success and return now an error makes a difference :)

Comment (anjiahao1) [reply sched/pthread/pthread_join.c:155]: > 现在返回成功和现在返回错误会有所不同:)

https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_join.html

Comment (pkarashchenko) [reply sched/pthread/pthread_join.c:155]: Seems like you refer to outdated version. Please use this one: https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_join.html

Screenshot 2022-04-07 at 10 06 44

Comment (anjiahao1) [reply sched/pthread/pthread_join.c:155]: > Seems like you refer to outdated version. Please use this one: https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_join.html
> 
> Screenshot 2022-04-07 at 10 06 44

yes，it also `If an implementation detects that the value specified by the thread argument to pthread_join() does not refer to a joinable thread, it is recommended that the function should fail and report an [EINVAL] error.`

Comment (pkarashchenko) [reply sched/pthread/pthread_join.c:155]: So we should add `else { ret = EINVAL }`?

File: sched/pthread/pthread_join.c [hunk 3]
```diff
@@ -138,6 +138,12 @@ int pthread_join(pthread_t thread, FAR pthread_addr_t *pexit_value)
     }
   else
     {
+      if (pjoin->detached)
+        {
+          leave_cancellation_point();
```
Comment (pkarashchenko) [inline sched/pthread/pthread_join.c:143]: `pthread_sem_give(&group->tg_joinsem);` is missing. The previous logic with `if` / `else` would perform clean-up properly

**Review by pkarashchenko — CHANGES_REQUESTED** [review]

**Review by pkarashchenko — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@anjiahao1 @Ouss4 
I noticed that the ostest failed with esp32-devkitc:smp (real board, not QEMU).
If I revert this PR locally, it works.

```
sem_waiter: Starting wait for condition
cancel_test: Canceling thread
cancel_test: Joining
cancel_test: PASS pthread_join failed with status=ESRCH
cancel_test: Test 5: Non-cancelable threads
cancel_test: Re-starting thread (non-cancelable)
restart_thread: Destroying cond
restart_thread: Destroying mutex
restart_thread: Re-starting thread
start_thread: Initializing mutex
start_thread: Initializing cond
start_thread: Starting thread
start_thread: Yielding
sem_waiter: Taking mutex
sem_waiter: Starting wait for condition
sem_waiter: Setting non-cancelable
cancel_test: Canceling thread
[CPU0] xtensa_panic: Unhandled Exception 2 task: pt-0x400e196c
[CPU0] xtensa_dumpstate: CPU0:
[CPU0] xtensa_registerdump:    PC: 4008010f    PS: 00040536
[CPU0] xtensa_registerdump:    A0: 00000002    A1: 3ffeb5a0    A2: 00000002    A3: 3ffe52dc
[CPU0] xtensa_registerdump:    A4: 3ffb01fc    A5: 20000000    A6: 3ffb0178    A7: 00000001
[CPU0] xtensa_registerdump:    A8: 3ffe52dc    A9: 3ffeb580   A10: 00000000   A11: 00000002
[CPU0] xtensa_registerdump:   A12: 3ffb0344   A13: 3ffb0348   A14: 00060222   A15: 00000000
[CPU0] xtensa_registerdump:   SAR: 0000001f CAUSE: 0000001d VADDR: ffffffd2
[CPU0] xtensa_registerdump:  LBEG: 00000000  LEND: 00000000  LCNT: 00000000
[CPU0] xtensa_dumpstate: sp:     3ffb0e10
[CPU0] xtensa_dumpstate: IRQ stack:
[CPU0] xtensa_dumpstate:   base: 3ffb0690
[CPU0] xtensa_dumpstate:   size: 00000800
[CPU0] xtensa_dumpstate:   used: 000002d0
[CPU0] xtensa_stackdump: 3ffb0e00: 800d57a6 3ffb0e50 00000000 3f4001db 00000000 3ffb0e70 00000002 3ffeb518
[CPU0] xtensa_stackdump: 3ffb0e20: 3ffe96d8 3ffb0e30 00000008 3ffb3018 3ffb1690 00001ff8 3ffb0690 3ffe5258
[CPU0] xtensa_stackdump: 3ffb0e40: 40080945 3ffb0e70 00000002 3ffeb518 3f40023e 00000002 3ffe53a0 00000001
[CPU0] xtensa_stackdump: 3ffb0e60: 00040023 3ffb0e90 00040536 00040026 3ffe52dc 3ffeb580 00000000 00000002
[CPU0] xtensa_dumpstate: sp:     3ffeb5a0
[CPU0] xtensa_dumpstate: User stack:
[CPU0] xtensa_dumpstate:   base: 3ffe96d8
[CPU0] xtensa_dumpstate:   size: 00001ff8
[CPU0] xtensa_dumpstate:   used: 00000248
[CPU0] xtensa_stackdump: 3ffeb5a0: 800de6b3 3ffeb5e0 ffffffff 3ffe5258 00000000 3ffeb5e0 ffffffff 3ffe5258
[CPU0] xtensa_stackdump: 3ffeb5c0: 3ffe70bc 00000001 00000000 00000000 800e6181 3ffeb610 ffffffff 00060220
[CPU0] xtensa_stackdump: 3ffeb5e0: ffffffff 00000000 00060422 00000084 3ffe96b8 00000000 00060022 00000084
[CPU0] xtensa_stackdump: 3ffeb600: 800e5c1c 3ffeb630 00001021 00000000 0000001f 3f4043c2 00000000 3ffb3fa6
[CPU0] xtensa_stackdump: 3ffeb620: 800e19e3 3ffeb650 00000000 00000000 3ffe5258 3ffe71a3 00000000 3ffb0330
[CPU0] xtensa_stackdump: 3ffeb640: 800de68c 3ffeb670 00000000 3ffb22dc 0000000c 0000001e 3ffb22d8 00000026
[CPU0] xtensa_stackdump: 3ffeb660: 800e4778 3ffeb690 400e196c 00000000 00060720 3ffe9160 00000000 00000000
[CPU0] xtensa_stackdump: 3ffeb680: 00000000 3ffeb6b0 3ffe5258 400de684 00000000 3ffe5258 00000001 00000000
[CPU0] xtensa_stackdump: 3ffeb6a0: 00000000 3ffeb6d0 00000000 00000000 00000001 3ffe70bc 00000000 00000000
[CPU0] xtensa_showtasks:    PID    PRI   CPU      USED     STACK   FILLED    COMMAND
[CPU0] xtensa_showtasks:   ----   ----  ----       784      2048    38.2%    irq
[CPU0] xtensa_dump_task:      0      0     0       560      3040    18.4%    CPU0 IDLE
[CPU0] xtensa_dump_task:      1      0     1       200      3040     6.5%    CPU1 IDLE
[CPU0] xtensa_dump_task:      2    100     0      1528      3032    50.3%    nsh_main
[CPU0] xtensa_dump_task:     38    100     0       584      8184     7.1%    pt-0x400e196c
[CPU0] xtensa_dump_task:     23    100     0      1096      2024    54.1%    ostest
[CPU0] xtensa_dump_task:     24    100     0      1112      8120    13.6%    ostest
```

**Comment by anjiahao1** [comment]
> @anjiahao1 @Ouss4 I noticed that the ostest failed with esp32-devkitc:smp (real board, not QEMU). If I revert this PR locally, it works.
> 
> ```
> sem_waiter: Starting wait for condition
> cancel_test: Canceling thread
> cancel_test: Joining
> cancel_test: PASS pthread_join failed with status=ESRCH
> cancel_test: Test 5: Non-cancelable threads
> cancel_test: Re-starting thread (non-cancelable)
> restart_thread: Destroying cond
> restart_thread: Destroying mutex
> restart_thread: Re-starting thread
> start_thread: Initializing mutex
> start_thread: Initializing cond
> start_thread: Starting thread
> start_thread: Yielding
> sem_waiter: Taking mutex
> sem_waiter: Starting wait for condition
> sem_waiter: Setting non-cancelable
> cancel_test: Canceling thread
> [CPU0] xtensa_panic: Unhandled Exception 2 task: pt-0x400e196c
> [CPU0] xtensa_dumpstate: CPU0:
> [CPU0] xtensa_registerdump:    PC: 4008010f    PS: 00040536
> [CPU0] xtensa_registerdump:    A0: 00000002    A1: 3ffeb5a0    A2: 00000002    A3: 3ffe52dc
> [CPU0] xtensa_registerdump:    A4: 3ffb01fc    A5: 20000000    A6: 3ffb0178    A7: 00000001
> [CPU0] xtensa_registerdump:    A8: 3ffe52dc    A9: 3ffeb580   A10: 00000000   A11: 00000002
> [CPU0] xtensa_registerdump:   A12: 3ffb0344   A13: 3ffb0348   A14: 00060222   A15: 00000000
> [CPU0] xtensa_registerdump:   SAR: 0000001f CAUSE: 0000001d VADDR: ffffffd2
> [CPU0] xtensa_registerdump:  LBEG: 00000000  LEND: 00000000  LCNT: 00000000
> [CPU0] xtensa_dumpstate: sp:     3ffb0e10
> [CPU0] xtensa_dumpstate: IRQ stack:
> [CPU0] xtensa_dumpstate:   base: 3ffb0690
> [CPU0] xtensa_dumpstate:   size: 00000800
> [CPU0] xtensa_dumpstate:   used: 000002d0
> [CPU0] xtensa_stackdump: 3ffb0e00: 800d57a6 3ffb0e50 00000000 3f4001db 00000000 3ffb0e70 00000002 3ffeb518
> [CPU0] xtensa_stackdump: 3ffb0e20: 3ffe96d8 3ffb0e30 00000008 3ffb3018 3ffb1690 00001ff8 3ffb0690 3ffe5258
> [CPU0] xtensa_stackdump: 3ffb0e40: 40080945 3ffb0e70 00000002 3ffeb518 3f40023e 00000002 3ffe53a0 00000001
> [CPU0] xtensa_stackdump: 3ffb0e60: 00040023 3ffb0e90 00040536 00040026 3ffe52dc 3ffeb580 00000000 00000002
> [CPU0] xtensa_dumpstate: sp:     3ffeb5a0
> [CPU0] xtensa_dumpstate: User stack:
> [CPU0] xtensa_dumpstate:   base: 3ffe96d8
> [CPU0] xtensa_dumpstate:   size: 00001ff8
> [CPU0] xtensa_dumpstate:   used: 00000248
> [CPU0] xtensa_stackdump: 3ffeb5a0: 800de6b3 3ffeb5e0 ffffffff 3ffe5258 00000000 3ffeb5e0 ffffffff 3ffe5258
> [CPU0] xtensa_stackdump: 3ffeb5c0: 3ffe70bc 00000001 00000000 00000000 800e6181 3ffeb610 ffffffff 00060220
> [CPU0] xtensa_stackdump: 3ffeb5e0: ffffffff 00000000 00060422 00000084 3ffe96b8 00000000 00060022 00000084
> [CPU0] xtensa_stackdump: 3ffeb600: 800e5c1c 3ffeb630 00001021 00000000 0000001f 3f4043c2 00000000 3ffb3fa6
> [CPU0] xtensa_stackdump: 3ffeb620: 800e19e3 3ffeb650 00000000 00000000 3ffe5258 3ffe71a3 00000000 3ffb0330
> [CPU0] xtensa_stackdump: 3ffeb640: 800de68c 3ffeb670 00000000 3ffb22dc 0000000c 0000001e 3ffb22d8 00000026
> [CPU0] xtensa_stackdump: 3ffeb660: 800e4778 3ffeb690 400e196c 00000000 00060720 3ffe9160 00000000 00000000
> [CPU0] xtensa_stackdump: 3ffeb680: 00000000 3ffeb6b0 3ffe5258 400de684 00000000 3ffe5258 00000001 00000000
> [CPU0] xtensa_stackdump: 3ffeb6a0: 00000000 3ffeb6d0 00000000 00000000 00000001 3ffe70bc 00000000 00000000
> [CPU0] xtensa_showtasks:    PID    PRI   CPU      USED     STACK   FILLED    COMMAND
> [CPU0] xtensa_showtasks:   ----   ----  ----       784      2048    38.2%    irq
> [CPU0] xtensa_dump_task:      0      0     0       560      3040    18.4%    CPU0 IDLE
> [CPU0] xtensa_dump_task:      1      0     1       200      3040     6.5%    CPU1 IDLE
> [CPU0] xtensa_dump_task:      2    100     0      1528      3032    50.3%    nsh_main
> [CPU0] xtensa_dump_task:     38    100     0       584      8184     7.1%    pt-0x400e196c
> [CPU0] xtensa_dump_task:     23    100     0      1096      2024    54.1%    ostest
> [CPU0] xtensa_dump_task:     24    100     0      1112      8120    13.6%    ostest
> ```

ok,i don't have esp32 borad,and i try this patch in rp2040 (real borad) with smp,it pass ostest,and sim pass too.i will buy a esp32 borad at this weekend,and try it. can you tell me,your esp32 chip version?

**Comment by masayuki2009** [comment]
> can you tell me,your esp32 chip version?

@anjiahao1 
The following boot messages show that the chip revision is 3

```
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:6604
ho 0 tail 12 room 4
load:0x40078000,len:14780
load:0x40080400,len:3792
entry 0x40080694
I (29) boot: ESP-IDF v4.4 2nd stage bootloader
I (29) boot: compile time 19:32:08
I (29) boot: chip revision: 3
I (32) boot_comm: chip revision: 3, min. bootloader chip revision: 0
I (39) boot.esp32: SPI Speed      : 40MHz
I (43) boot.esp32: SPI Mode       : DIO
I (48) boot.esp32: SPI Flash Size : 8MB
I (53) boot: Enabling RNG early entropy source...
I (58) boot: Partition Table:
I (62) boot: ## Label            Usage          Type ST Offset   Length
I (69) boot:  0 factory          factory app      00 00 00010000 00100000
I (76) boot: End of partition table
I (80) boot_comm: chip revision: 3, min. application chip revision: 0
I (88) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=0863ch ( 34364) map
I (109) esp_image: segment 1: paddr=00018664 vaddr=3ffb3d00 size=0010ch (   268) load
I (109) esp_image: segment 2: paddr=00018778 vaddr=40080000 size=01cdch (  7388) load
I (118) esp_image: segment 3: paddr=0001a45c vaddr=00000000 size=05bbch ( 23484) 
I (132) esp_image: segment 4: paddr=00020020 vaddr=400d0020 size=16cb8h ( 93368) map
I (167) boot: Loaded app from partition at offset 0x10000
I (167) boot: Disabling RNG early entropy source...
?
NuttShell (NSH) NuttX-10.3.0-RC0
nsh> uname -a
NuttX 10.3.0-RC0 8d66dbc068 Apr  8 2022 21:15:42 xtensa esp32-devkitc
```

**Comment by anjiahao1** [comment]
@masayuki2009 hi，i bought a esp32devkitc, and tried `ostest` in `smp` ,it pass in latest code.
```
barrier_test: Thread 2 completed with result=0                                                                        barrier_test: Thread 3 completed with result=0                                                                        barrier_func: Thread 4 done                                                                                           barrier_func: Thread 5 done                                                                                           barrier_func: Thread 6 done                                                                                           barrier_test: Thread 4 completed with result=0                                                                        barrier_test: Thread 5 completed with result=0                                                                        barrier_test: Thread 6 completed with result=0                                                                        barrier_test: Thread 7 completed with result=0                                                                                                                                                                                              End of test memory usage:                                                                                             VARIABLE  BEFORE   AFTER                                                                                              ======== ======== ========                                                                                            arena       4c1f0    4c1f0                                                                                            ordblks         7        8                                                                                            mxordblk    2b790    2b790                                                                                            uordblks     8840     cbc0                                                                                            fordblks    439b0    3f630                                                                                                                                                                                                                  Final memory usage:                                                                                                   VARIABLE  BEFORE   AFTER                                                                                              ======== ======== ========                                                                                            arena       4c1f0    4c1f0                                                                                            ordblks         4        8                                                                                            mxordblk    2b790    2b790                                                                                            uordblks     5040     cbc0                                                                                            fordblks    471b0    3f630                                                                                            user_main: Exiting                                                                                                    ostest_main: Exiting with status 0                                                                                    nsh> 
```

**Comment by masayuki2009** [comment]
>@masayuki2009 hi，i bought a esp32devkitc, and tried ostest in smp ,it pass in latest code.

@anjiahao1 
OK. By the way, which version of the compiler are you using now?

```
$ xtensa-esp32-elf-gcc --version
xtensa-esp32-elf-gcc (crosstool-NG esp-2021r2-patch3) 8.4.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by anjiahao1** [comment]
> > @masayuki2009 hi，i bought a esp32devkitc, and tried ostest in smp ,it pass in latest code.
> 
> @anjiahao1 OK. By the way, which version of the compiler are you using now?
> 
> ```
> $ xtensa-esp32-elf-gcc --version
> xtensa-esp32-elf-gcc (crosstool-NG esp-2021r2-patch3) 8.4.0
> Copyright (C) 2018 Free Software Foundation, Inc.
> This is free software; see the source for copying conditions.  There is NO
> warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
> ```

same version as you
```
$ xtensa-esp32-elf-gcc --version
xtensa-esp32-elf-gcc (crosstool-NG esp-2021r2-patch3) 8.4.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by masayuki2009** [comment]
@anjiahao1 

It seems that https://github.com/apache/incubator-nuttx/pull/6056 fixed the issue.
I confirmed that the latest upstream works.

**Comment by xiaoxiang781216** [comment]
@masayuki2009 could you help try this patch(https://github.com/apache/incubator-nuttx/pull/6063) related to stack colorization? which fix the problem we found when ASAN eanble.
